net: buf: keep memory alignment provided by k_heap_alloc and k_malloc

jfischer-no opened the pull request:
Allocation callback in net_buf_heap_cb and net_buf_var_cb
used for net_bufs with variable size payloads, defined by
NET_BUF_POOL_HEAP_DEFINE or NET_BUF_POOL_VAR_DEFINE,
allocate one more octet for internal variable ref_count,
used in function generic_data_ref(), which in turn is needed
for net_buf_clone()).
The user gets a buffer which is shifted by one octet in memory
block. This breaks alignment provided k_heap_alloc and k_malloc.

jfischer-no committed:
net: buf: keep memory alignment provided by k_heap_alloc and k_malloc

Allocation callback in net_buf_heap_cb and net_buf_var_cb
used for net_bufs with variable size payloads, defined by
NET_BUF_POOL_HEAP_DEFINE or NET_BUF_POOL_VAR_DEFINE,
allocate one more octet for internal variable ref_count,
used in function generic_data_ref(), which in turn is needed
for net_buf_clone()).
The user gets a buffer which is shifted by one octet in memory
block. This breaks alignment provided k_heap_alloc and k_malloc.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jhedberg commented:
I’m not really opposed to this, curious to hear where you benefit from it considering that in almost all cases the buffer payload is some protocol data that’s not assumed to be properly aligned, ie we access it using packed structs or using the various net buf data getter functions that don’t assume any specific alignment for the payload.

jfischer-no commented:
> I’m not really opposed to this, curious to hear where you benefit from it considering that in almost all cases the buffer payload is some protocol data that’s not assumed to be properly aligned

I use it in USB device controller API, some controllers allow to use data for transfers directly from RAM without storing it in specific area or FIFO, but data block must be aligned. IMHO it is a kind of bug because it does not keep the alignment provided by k_heap what one could assume. As I see fixed size pools can not provide aligned buffers as well. I could possibly solve it differently by definition of own callbacks and NET_BUF_POOL_UDC_DEFINE macro.

carlescufi reviewed (changes requested):
Please update the documentation for `k_heap_alloc` stating that it aligns the memory to `void *`.

jfischer-no committed:
include: kernel: add comment about alignmend of allocated memory

Add comment that memory block allocated by k_malloc() and
k_heap_alloc(), are aligned to multiple of pointer sizes.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
@rlubos @jhedberg @tbursztyka could you please review?

M1cha commented:
Am I right in that this messes up `generic_data_ref`? That one still uses `1` insead of `sizeof(void*)`:
```c
ref_count = data - 1;
```

If true, ref and unref would use different memory locations. I'm not sure how that would not be caught by unit and manual tests though since it would cause memory leaks all over the place.

jhedberg commented:
@jfischer-no could you respond to the comment from @M1cha above? Is there some place in the code that you forgot to update?

jfischer-no commented:
> Am I right in that this messes up `generic_data_ref`? That one still uses `1` insead of `sizeof(void*)`:
> 
> ```c
> ref_count = data - 1;
> ```
> 
> If true, ref and unref would use different memory locations. I'm not sure how that would not be caught by unit and manual tests though since it would cause issues all over the place.

It looks like I screwed up on this one, cloned net_buf data would point to a data block that has already been freed:

https://github.com/zephyrproject-rtos/zephyr/blob/fa5b706871f01bb3bba3b2d9f774d27615855e7e/subsys/net/buf.c#L96-L103

There are net_buf reference count (member ref of struct net_buf) and data reference count at the top of allocated data. This PR is about data reference count.
Probably no one caught it because it is only used for NET_BUF_POOL_VAR_DEFINE() and  NET_BUF_POOL_HEAP_DEFINE() to increase the data reference count in `net_buf_clone()`. NET_BUF_POOL_HEAP_DEFINE is not used in the tree. 

test_net_buf_clone() would not catch it, @jhedberg any ideas other than the one-line fix below?

```diff
diff --git a/subsys/net/buf.c b/subsys/net/buf.c
index dc6f5dc8e0d..462a21198e2 100644
--- a/subsys/net/buf.c
+++ b/subsys/net/buf.c
@@ -93,7 +93,7 @@ static uint8_t *generic_data_ref(struct net_buf *buf, uint8_t *data)
 {
        uint8_t *ref_count;
 
-       ref_count = data - 1;
+       ref_count = data - sizeof(void *);
        (*ref_count)++;
 
        return data;
```
